faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -112,57 +127,80 @@ public Builder withMetrics(MetricRegistry registry) {
-        r = new MetricRegistry();
+        registry = new MetricRegistry();
       else
-        r = metricRegistry;
+        registry = metricRegistry;
 
-      AsyncHttpClient c;
+      AsyncHttpClient httpClient;
       if (client == null) {
-        AsyncHttpClientConfig config = new DefaultAsyncHttpClientConfig.Builder()
-          .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS)
-          .setRequestTimeout(DEFAULT_REQUEST_TIMEOUT_MS)
-          .setPooledConnectionIdleTimeout(DEFAULT_IDLE_TIMEOUT_MS)
-          .setMaxRequestRetry(0)
-          .build();
-        c = new DefaultAsyncHttpClient(config);
-      } else
-        c = client;
+        httpClient = new DefaultAsyncHttpClient(
+          new DefaultAsyncHttpClientConfig.Builder()
+            .setConnectTimeout(DEFAULT_CONNECTION_TIMEOUT_MS)
+            .setRequestTimeout(DEFAULT_REQUEST_TIMEOUT_MS)
+            .setPooledConnectionIdleTimeout(DEFAULT_IDLE_TIMEOUT_MS)
+            .setMaxRequestRetry(DEFAULT_MAX_RETRIES)
+            .build()
+        );
+      } else {
+        httpClient = client;
+      }
 
       URL root;
       if (faunaRoot == null)
-        root = new URL("https://rest.faunadb.com");
+        root = FAUNA_ROOT;
       else
         root = faunaRoot;
 
-      return new Connection(root, authToken, c, r);
+      return new Connection(root, authToken, httpClient, registry, new AtomicInteger(1));
     }
   }
 
-  private static String XFaunaDBHost = "X-FaunaDB-Host";
-  private static String XFaunaDBBuild = "X-FaunaDB-Build";
+  private static final String ASCII = "ASCII";
+  private static final String X_FAUNADB_HOST = "X-FaunaDB-Host";
+  private static final String X_FAUNADB_BUILD = "X-FaunaDB-Build";
 
   private final URL faunaRoot;
-  private final String authToken;
   private final String authHeader;
   private final AsyncHttpClient client;
   private final MetricRegistry registry;
+  private final AtomicInteger refCount;
 
   private final Logger log = LoggerFactory.getLogger(getClass());
   private final ObjectMapper json = new ObjectMapper();
 
-  Connection(URL faunaRoot, String authToken, AsyncHttpClient client, MetricRegistry registry) throws UnsupportedEncodingException {
+  private Connection(URL faunaRoot, String authToken, AsyncHttpClient client, MetricRegistry registry, AtomicInteger refCount) {
     this.faunaRoot = faunaRoot;
-    this.authToken = authToken;
     this.authHeader = generateAuthHeader(authToken);
     this.client = client;
     this.registry = registry;
+    this.refCount = refCount;
+  }
+
+  /**
+   * Creates a new short live connection sharing its underneath {@link AsyncHttpClient}. Queries submited to a
+   * session connection will be authenticated with the token provided.
+   *
+   * @param authToken token that will be used to authenticate requests
+   * @return a new {@link Connection}
+   */
+  public Connection newSessionConnection(String authToken) {
+    if (refCount.incrementAndGet() > 1)
```

Comment:
Java can be so ugly :( This would look so much nicer in Java8. Alas.

It's probably better to allow customers the ability to save the new client. After thinking about it more, I realized that Java-land is more procedural and forcing customers to encapsulate the query in a function is most likely going to annoy them. 

I still think it's clearer to bump the refcount only when the connection is used, which is any time it's passed to a FaunaClient ctor.

The way refCount is passed around still feels wrong to me. I think, ideally, the http client would have an `retain/release` interface since, ultimately, that's what we're trying to protect. Like I said, it's mostly about principle. What you have now is fine, though.